java/ql/src/semmle/code/java/dataflow/FlowSources.qll

```diff
@@ -270,3 +271,39 @@ class AndroidIntentInput extends DataFlow::Node {
     )
   }
 }
+
+/**
+ * Method access to external inputs of `android.content.Intent` object
+ */
+class IntentGetExtraMethodAccess extends MethodAccess {
+  IntentGetExtraMethodAccess() {
+    exists(AndroidComponent ac |
+      this.getEnclosingCallable().getDeclaringType() = ac and ac.isExported()
+    ) and
+    (
+      this.getMethod().getName().regexpMatch("get\\w+Extra") and
+      this.getMethod().getDeclaringType() instanceof TypeIntent
+      or
+      this.getMethod().getName().regexpMatch("get\\w+") and
+      this.getQualifier().(MethodAccess).getMethod().hasName("getExtras") and
+      this.getQualifier().(MethodAccess).getMethod().getDeclaringType() instanceof TypeIntent
+    )
+  }
+}
+
+/**
+ * Android intent extra source
+ */
+private class AndroidIntentExtraSource extends RemoteFlowSource {
```

Comment:
I don't know the specific definitions for what exactly is considered to be a `RemoteFlowSource` and what a `LocalUserInput`.
But aren't intents more like local user input or is there some way to remotely trigger intents?

Reply:
To my understanding, a `LocalUserInput` is more like configuration parameters in the application or OS, which can be controlled by the local user process running the app, e.g. resource or configuration property. 

I noticed there is an implementation of `AndroidIntentInput` in the same file `FlowSources.qll` already. However, I cannot find any reference to it in the CodeQL repository so I'm not sure what's its intended usage.

Android intents can be used both within the same application or across application boundaries. I think remotely triggered intents (the latter) are of particular interests since they can be controlled by malicious apps/users thus are more risky in nature. An Android application with many activities and intents passing data internally only is not vulnerable since malicious users/apps cannot affect its data processing. This is the reason I created `AndroidIntentExtraSource` to detect remote data used in exposed intents.